### PR TITLE
Proposed fix for issue #21519: Reshape layer does not handle -1 shape…

### PR DESCRIPTION
**Pull request by roebel**
I have implemented a fix that avoids precalculating a statically resolved reshape target in the build method and instead properly calculates the proper shape information for the problematic case when the shape information is not complete in the call method. The method appears to pass the reshape tests for all backends.

### CODE REVIEW

File: keras/src/layers/reshaping/reshape.py [hunk 1]
```diff
@@ -53,19 +65,23 @@ def compute_output_spec(self, inputs):
...
     def call(self, inputs):
+        target_shape = self.target_shape
+        if self.need_explicit_shape_for_batch_size_None and (inputs.shape[0] is None):
+            input_nonbatch_shape = tuple(inputs.shape[1:])
+            if input_nonbatch_shape.count(None) == 0:
+                # If the input shape is fully defined, we can compute the desired target_shape
+                if True:
+                    inp_nonbatch_size = math.prod(inputs.shape[1:])
+                else:
+                    inp_nonbatch_size = ops.prod(ops.shape(inputs)[1:])
```
[the remark this answers is not in the document]
Comment (roebel) [reply keras/src/layers/reshaping/reshape.py]: Yes indeed, I wondered in fact whether a change like this:

```
    def call(self, inputs):
        target_shape = self.target_shape
        if self.need_explicit_shape_for_batch_size_None and (inputs.shape[0] is None):
                inp_nonbatch_size = ops.prod(ops.shape(inputs)[1:])
                target_shape = tuple(d if d != -1 else (inp_nonbatch_size // self.new_size_no_minus_one) for d in self.target_shape)

        return ops.reshape(
            inputs, (ops.shape(inputs)[0],) + target_shape
        )
```

would have any advantage. Would it?

**Comment by roebel** [comment]
You find the test case in the new commit. For the context: When using convolutional models with time series data (voice processing), the time dimension is hardly ever known beforehand. You can probably force a dynamic dimension here, but for convolutional layers this is not required, and for the reshape layer it is not needed either.

> Instead redoing the work of calculating the dynamically resolved shape, it just lets ops.reshape handle it. No need to duplicated that work

I may not fully understand this, but I thought that the reshape layer has an advantage compared to the reshape operator, because it knows that the batch dimension stays the same. I wondered in fact why the test case: test_reshape_with_dynamic_batch_size_and_minus_one produces shape (None, 3, 8) and not (None, None, 8). As for example this does:

```
ops.reshape(keras.KerasTensor((None, 6, 4)), (None, -1, 8))
Out[8]: <KerasTensor shape=(None, None, 8), dtype=float32, sparse=False, ragged=False, name=keras_tensor_3>
```

**Comment by roebel** [comment]
I don't see any reason why the torch test fails. It runs fine here in my Keras development environment, and it also runs fine with the mini script I shared in the issue.

File: keras/src/layers/reshaping/reshape.py [hunk 2]
```diff
@@ -53,18 +69,22 @@ def compute_output_spec(self, inputs):
...
-        )
-        self._resolved_target_shape = tuple(
-            -1 if d is None else d for d in sample_output_shape
-        )
-
     def call(self, inputs):
-        return ops.reshape(
-            inputs, (ops.shape(inputs)[0],) + self._resolved_target_shape
-        )
+        target_shape = self.target_shape
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape.py]: I think it's easier to just transfer the code from `build` here:

```python
    def call(self, inputs):
        potentially_resolved_target_shape = (
            operation_utils.compute_reshape_output_shape(
                tuple(inputs.shape)[1:], self.target_shape, "target_shape"
            )
        )
        potentially_resolved_target_shape = tuple(
            -1 if d is None else d for d in potentially_resolved_target_shape
        )
        return ops.reshape(
            inputs, (ops.shape(inputs)[0],) + potentially_resolved_target_shape
        )
```

- you don't have to reimplement the computation of the missing dimension
- you don't have to deal with errors if the number of values is not divisible by `self.new_size_no_minus_one` (right now, that check is missing)
- you don't need `self.need_explicit_shape_for_batch_size_None` and `self.new_size_no_minus_one`

File: keras/src/layers/reshaping/reshape.py [hunk 3]
```diff
@@ -37,7 +39,21 @@ class Reshape(Layer):
...
+                "`-1` value. Received: target_shape={}".format(target_shape)
+            )
+        self.target_shape = target_shape
+        # precalculate all values that might be required
+        self.need_explicit_shape_for_batch_size_None = (
+            target_shape.count(-1) == 1
+        )
+        self.new_size_no_minus_one = math.prod(
+            d for d in target_shape if d != -1
+        )
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape.py]: Because you removed the `build` method, add `self.built = True` at the end of `__init__`.

File: keras/src/layers/reshaping/reshape_test.py [hunk 4]
```diff
@@ -100,6 +102,32 @@ def test_reshape_with_dynamic_batch_size_and_minus_one(self):
         reshaped = backend.compute_output_spec(layer.__call__, input)
         self.assertEqual(reshaped.shape, (None, 3, 8))
 
+    def test_reshape_layer_with_varying_input_size_and_minus_one(self):
+        input = KerasTensor((None, 6, 4))
+        layer = layers.Reshape((-1, 8))
+        layer.build(input.shape)
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape_test.py]: You could remove all the calls to `build` in this test and others, it's no longer needed (just to make the code shorter).

File: keras/src/layers/reshaping/reshape_test.py [hunk 5]
```diff
@@ -100,6 +102,32 @@ def test_reshape_with_dynamic_batch_size_and_minus_one(self):
         reshaped = backend.compute_output_spec(layer.__call__, input)
         self.assertEqual(reshaped.shape, (None, 3, 8))
 
+    def test_reshape_layer_with_varying_input_size_and_minus_one(self):
+        input = KerasTensor((None, 6, 4))
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape_test.py]: `input = KerasTensor((None, 6, 4))` is no longer needed in this test, remove.

File: keras/src/layers/reshaping/reshape_test.py [hunk 6]
```diff
@@ -100,6 +102,32 @@ def test_reshape_with_dynamic_batch_size_and_minus_one(self):
...
+        layer = layers.Reshape((-1, 8))
+        layer.build(input.shape)
+        res = layer(ops.ones((1, 6, 4), dtype="float32"))
+        self.assertEqual(res.shape, (1, 3, 8))
+        res = layer(ops.ones((1, 10, 4), dtype="float32"))
+        self.assertEqual(res.shape, (1, 5, 8))
+
+    def test_custom_reshape_model_with_varying_input_size_and_minus_one(self):
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape_test.py]: Let's do a compiled example instead, which exercise a different use case than the example right before. Also, the model can be just the `Reshape` layer:

```python
    def test_reshape_model_fit_with_varying_input_size_and_minus_one(self):
        def generator():
            yield (
                ops.ones((1, 12, 2), dtype="float32"),
                ops.zeros((1, 3, 8), dtype="float32"),
            )
            yield (
                ops.ones((1, 20, 2), dtype="float32"),
                ops.zeros((1, 5, 8), dtype="float32"),
            )

        layer = layers.Reshape((-1, 8))
        model = models.Sequential([layer])
        model.compile(loss="mean_squared_error")
        model.fit(generator())
```

File: keras/src/layers/reshaping/reshape.py [hunk 7]
```diff
@@ -37,7 +39,21 @@ class Reshape(Layer):
 
     def __init__(self, target_shape, **kwargs):
         super().__init__(**kwargs)
-        self.target_shape = tuple(target_shape)
+        target_shape = tuple(target_shape)
+        # test validity of target_shape
+        if target_shape.count(-1) > 1:
+            raise ValueError(
+                "The `target_shape` argument must not contain more than one "
+                "`-1` value. Received: target_shape={}".format(target_shape)
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape.py]: Please use an f-string for this.

**Comment by roebel** [comment]
Thanks for all the comments and suggestions.

File: keras/src/layers/reshaping/reshape.py [hunk 8]
```diff
@@ -11,11 +11,13 @@ class Reshape(Layer):
 
     Args:
         target_shape: Target shape. Tuple of integers, does not include the
-            samples dimension (batch size).
+            samples dimension (batch size). One element of the target_shape can
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape.py]: Nitpick: `` `target_shape` `` with backquotes.

File: keras/src/layers/reshaping/reshape.py [hunk 9]
```diff
@@ -11,11 +11,13 @@ class Reshape(Layer):
...
+            be -1 in which case the missing value is inferred from the length
+            of the array and remaining dimensions.
 
     Input shape:
-        Arbitrary, although all dimensions in the input shape must be
-        known/fixed. Use the keyword argument `input_shape` (tuple of integers,
+        Arbitrary, but required to be compatible with target shape.
+        Use the keyword argument `input_shape` (tuple of integers,
```
Comment (hertschuh) [inline keras/src/layers/reshaping/reshape.py]: >  Use the keyword argument `input_shape` ...

Now I'm noticing this documentation. This is actually a deprecated way of doing this. Plus, it's not needed.

Just remove that last part.

**Review by hertschuh — APPROVED** [review]
Thanks a lot for the PR!
